atomicapp/nulecule/main.py

```diff
@@ -185,14 +158,55 @@ def init(app_name, destination=None, app_version='1.0',
                     os.remove(templ_path)
 
             # Copy rendered templates to destination directory
-            distutils.dir_util.copy_tree(tmpdir, destination, True)
+            copy_tree(tmpdir, destination, True)
         finally:
             # Remove temporary working directory
-            distutils.dir_util.remove_tree(tmpdir)
+            remove_tree(tmpdir)
         return destination
 
-    def unpack(self, update=False,
-               dryrun=False, nodeps=False, config=None):
+    def digraph(self, update=False, dryrun=False, nodeps=False, config=None,
+                **kwargs):
+        """
+        Args:
+            update (bool): Update existing Nulecule application in
+                           app_path, if True
+            dryrun (bool): Do not make any change to the host system
+            nodeps (bool): Do not unpack external dependencies
+            config (dict): Config data, if any, to use for unpacking
+            kwargs (dict): Extra keyword arguments
+
+        Returns:
+            networkx.DiGraph of Nulecule objects
+        """
+        digraph = DiGraph()
+
+        (progenitor,
+         internals,
+         externals) = self.unpack(config=config, dryrun=dryrun, nodeps=nodeps,
+                                  update=update, **kwargs)
+
+        digraph.add_node(progenitor)
+
+        for internal in internals:
+            digraph.add_edge(progenitor, internal)
+
+        for external in externals:
+            app_spec = Utils.getSourceImage(external)
+            name = external[NAME_KEY]
+            destination = "%s/%s/%s" % (self.app_path, EXTERNAL_APP_DIR, name)
+            if app_spec is not None:
+                kwargs.update({'namespace': name})
+                kwargs.update(self._process_app_spec(app_spec, destination))
+                subgraph = self.digraph(**kwargs)
+                if subgraph.order() > 0:
+                    patronage = internals or [progenitor]
```

Comment:
Of all the lines of code, you found the one that I thought was most worthy of some commentary, yet hardest to explain. Essentially what's going on here is as follows, which probably makes it worthwhile to separate things back into two classes, assuming they don't inherit from the same base class.

```progenitor``` and ```externals``` each have the ```progenitor``` attribute set to ```true``` and align with what was previously the ```Nulecule``` class, whereas ```internals``` describe the actual units of deployment and align with what was previously the ```NuleculeComponent``` class. It seems reasonable to restore that basic class hierarchy.

Take for example two use cases.

```
---
specversion: 0.0.2
id: NuleculeA

metadata:
  name: NuleculeA
  appversion: 1.0
  description: App description

graph:
  - name: NuleculeB
    source: docker://NuleculeB

  - name: NuleculeC
    source: docker://NuleculeC
```

I'll revert to using the prior lingo to explain what's going on. In this first use case, we have a ```Nulecule``` with relationships only to other ```Nulecule``` type objects, whereas in the latter case the ```Nulecule``` has a relationship to both ```Nulecule``` and ```NuleculeComponent``` type objects.

In the first example, since there are no ```NuleculeComponent``` type objects, we need to create an edge from ```NuleculeA``` (i.e. ```or [progenitor]```) to ```NuleculeB``` and ```NuleculeC```.

```
     NuleculeA
        / \
NuleculeB NuleculeC
```

```
---
specversion: 0.0.2
id: NuleculeA

metadata:
  name: NuleculeA
  appversion: 1.0
  description: App description

graph:
  - name: NuleculeComponent1
    artifacts:
      kubernetes:
        - file://artifacts/kubernetes/template.yaml

  - name: NuleculeComponent2
    artifacts:
      kubernetes:
        - file://artifacts/kubernetes/template.yaml

  - name: NuleculeB
    source: docker://NuleculeB

  - name: NuleculeC
    source: docker://NuleculeC
```

In this second use case, ```NuleculeA``` has a first-order relationship with ```NuleculeComponent1``` and ```NuleculeComponent2``` and a second-order relationship with ```NuleculeB``` and ```NuleculeC```, since either or both of ```NuleculeComponent1``` and ```NuleculeComponent2``` (i.e. ```internals```) are what actually depend on ```NuleculeB``` or ```NuleculeC```.

Given the inability to map which of ```NuleculeComponent1``` or ```NuleculeComponent2``` actually depends on ```NuleculeB``` or ```NuleculeC``` with any precision, more sweeping assumptions must be made about their relationships. That is why I was originally allured by the assumption that there was a one-to-one relationship between ```Nulecule``` and ```NuleculeComponent```, as opposed to one-to-many, since some granularity is lost in the latter case.

```
              NuleculeA
                 / \
 NuleculeComponent1 NuleculeComponent2
        / \                 / \
NuleculeB NuleculeC NuleculeB NuleculeC
```